### PR TITLE
Feature/useful join

### PR DESCRIPTION
**Pull request by bilow-factual**
Hey, I made `join` work sort of, I think, can you take a look?

My grand finale test that passed was:

`./ni 1p'cart ["x", "x", "y", "z"], ["foo", "baz", "bar"]' j[ i[x yo] i[z co] i[z bo] ]`

### CODE REVIEW

File: core/row/join.pl [hunk 1]
```diff
@@ -8,22 +8,44 @@
...
-        $reof ||= !defined $rrow until $reof or $rkey ge $lkey;
-      chomp($lkey = join "\t", (split /\t/, $lrow = <STDIN>, $rlimit + 1)[@rcols]),
-        $leof ||= !defined $lrow until $leof or $lkey ge $rkey;
-    }
+  chomp(my $lkey = join "\t", (split /\t/, my $lrow = <STDIN>, $llimit + 1)[@lcols]);
+  chomp(my $rkey = join "\t", (split /\t/, my $rrow = <$fh>,   $rlimit + 1)[@rcols]);
+
+  while (!$leof && !$reof) {
+    if ($lkey lt $rkey) {
+      print "$lkey < $rkey\n";
```
Comment (spencertipping) [inline core/row/join.pl]: uh oh, debugging stuff

File: core/row/join.pl [hunk 2]
```diff
@@ -8,22 +8,44 @@
...
+      chomp($lkey = join "\t", (split /\t/, $lrow = <STDIN>, $llimit + 1)[@lcols]);
+      $leof ||= !defined $lrow;
+    } elsif ($lkey gt $rkey) {
+      print "$lkey > $rkey\n";
+      chomp($rkey = join "\t", (split /\t/, $rrow = <$fh>,   $rlimit + 1)[@rcols]);
+      $reof ||= !defined $rrow;
+    } else {
+      @rrows = ($rrow,);
```
Comment (spencertipping) [inline core/row/join.pl]: this will work - you can also say `@rrows = $rrow` and perl will do the right thing

File: core/row/join.pl [hunk 3]
```diff
@@ -8,22 +8,44 @@
...
+      }
 
-    if ($lkey eq $rkey and !$leof && !$reof) {
-      chomp $lrow;
-      print "$lrow\t$rrow";
+      while(!$leof) {
+        chomp $lrow;
+        print "$lrow\t$_" for @rrows;
+        chomp(my $new_lkey = join "\t", (split /\t/, $lrow = <STDIN>, $llimit + 1)[@lcols]);
+        if ($new_lkey eq $lkey) {
```
Comment (spencertipping) [inline core/row/join.pl]: do you want to say `unless()` here?

File: doc/cheatsheet_perl.md [hunk 4]
```diff
@@ -30,6 +16,57 @@ Perl variables are indexed with "sigils" to signify their type. This can be
...
+* If you try to operate on an undefined value, it will define itself to be compatible with the operation you use and default to zero, the empty string, the empty array/hash, etc.
+
+
+### Lexical scoping
+
+Perl is lexically scoped; in general, when using `ni`, you will want to prefix every variable you define within a Perl mapper as scoped to the block, using the keyword `my`, as in `$ ni n10 p'my $x = 0; r $x'`
+
+The exception to this rule is for variables defined within a begin block. In this case `my` is not used at all. See: `$ ni n10 p'^{$x = 0} $x += 1; r $x'`.
+
+### Refernce Basics
```
Comment (spencertipping) [inline doc/cheatsheet_perl.md]: tyop

File: doc/cheatsheet_perl.md [hunk 5]
```diff
@@ -86,6 +123,11 @@ If you are unfamiliar with these functions, they are explained more extensively
 * `$<v> =~ s/regex//` -- substitution
 * `$<v> = tr/regex//` -- transliteration (also can be done with `y/regex//`)
 
+### Logical and Existence Operators
+* `exists` -- truthy when a key exists in a hash, falsey otherwise
+* `defined` -- truthy when a value is not `undef`, falsey otherwise
+* `&&`, `||` -- high priority boolean operators; use to perform boolean operations directly on scalar values
```
Comment (spencertipping) [inline doc/cheatsheet_perl.md:129]: this might be a good place to mention `//`; only the oldest perls don't support it

**Review by spencertipping — COMMENTED** [review]
this is great! some comments for ya

**Comment by bilow-factual** [comment]
💥 💥 💥

**Pull request merged**